File: server/internal/accounts/db/simplefin_tokens.go
package accountsdb

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/samber/lo"

	"tallyo/internal/accounts"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
)

const DefaultSimpleFinSyncCron = "0 6,18 * * *"

func (s *Store) CreateSimpleFinAccessToken(ctx context.Context, accessURL string, ownerID int64, label string) (*model.SimpleFinAccessToken, error) {
	nextSyncAt := time.Now().UTC()
	row, err := s.q.CreateSimpleFinAccessToken(ctx, dbgen.CreateSimpleFinAccessTokenParams{
		AccessUrl:  accessURL,
		Label:      lo.EmptyableToPtr(label),
		OwnerID:    ownerID,
		SyncCron:   DefaultSimpleFinSyncCron,
		NextSyncAt: &nextSyncAt,
	})
	if err != nil {
		return nil, fmt.Errorf("create simplefin access token: %w", err)
	}
	return simpleFinAccessTokenFromCreateRow(row), nil
}

func simpleFinAccessTokenFromCreateRow(row dbgen.CreateSimpleFinAccessTokenRow) *model.SimpleFinAccessToken {
	return &model.SimpleFinAccessToken{
		ID:           model.New(model.GlobalIDSimpleFinAccessToken, row.ID),
		Label:        row.Label,
		Owner:        &model.Owner{ID: model.New(model.GlobalIDOwner, row.OwnerID)},
		SyncCron:     row.SyncCron,
		LastSyncedAt: row.LastSyncedAt,
		NextSyncAt:   row.NextSyncAt,
		CreatedAt:    row.CreatedAt,
	}
}

func (s *Store) SimpleFinAccessTokens(ctx context.Context) ([]*model.SimpleFinAccessToken, error) {
	rows, err := s.q.SimpleFinAccessTokens(ctx, dbgen.SimpleFinAccessTokensParams{})
	return dbutil.MapRows(rows, err, simpleFinAccessTokenFromSQLRow)
}

func (s *Store) SimpleFinAccessTokenByID(ctx context.Context, id int64) (*model.SimpleFinAccessToken, error) {
	rows, err := s.q.SimpleFinAccessTokens(ctx, dbgen.SimpleFinAccessTokensParams{ID: &id})
	return dbutil.MapFirstRow(rows, err, simpleFinAccessTokenFromSQLRow)
}

func simpleFinAccessTokenFromSQLRow(row dbgen.SimpleFinAccessTokensRow) *model.SimpleFinAccessToken {
	return &model.SimpleFinAccessToken{
		ID:           model.New(model.GlobalIDSimpleFinAccessToken, row.ID),
		Label:        row.Label,
		Owner:        &model.Owner{ID: model.New(model.GlobalIDOwner, row.OwnerID), Name: row.OwnerName},
		SyncCron:     row.SyncCron,
		LastSyncedAt: row.LastSyncedAt,
		NextSyncAt:   row.NextSyncAt,
		CreatedAt:    row.CreatedAt,
	}
}

func simpleFinSecretFromRow(row dbgen.SimplefinAccessToken) accounts.SimpleFinAccessTokenSecret {
	return accounts.SimpleFinAccessTokenSecret{
		ID:                row.ID,
		AccessURL:         row.AccessUrl,
		OwnerID:           row.OwnerID,
		SyncCron:          row.SyncCron,
		LastSyncedAt:      row.LastSyncedAt,
		NextSyncAt:        row.NextSyncAt,
		NextBalanceSyncAt: row.NextBalanceSyncAt,
	}
}

func SimpleFinTokenSecretByConnID(ctx context.Context, q *dbgen.Queries, connID int64) (*accounts.SimpleFinAccessTokenSecret, error) {
	row, err := q.SimpleFinTokenSecretByConnID(ctx, dbgen.SimpleFinTokenSecretByConnIDParams{ConnID: connID})
	toSecret := func(row dbgen.SimplefinAccessToken) *accounts.SimpleFinAccessTokenSecret {
		secret := simpleFinSecretFromRow(row)
		return &secret
	}
	return dbutil.MapRow(row, err, toSecret)
}

func (s *Store) SimpleFinTokenSecretByConnID(ctx context.Context, connID int64) (*accounts.SimpleFinAccessTokenSecret, error) {
	return SimpleFinTokenSecretByConnID(ctx, s.q, connID)
}

func SimpleFinTokenSecretsDue(
	ctx context.Context,
	q *dbgen.Queries,
	kind accounts.SimpleFinTokenSecretKind,
	now time.Time,
) ([]accounts.SimpleFinAccessTokenSecret, error) {
	nowUTC := now.UTC()
	rows, err := q.SimpleFinTokenSecrets(ctx, dbgen.SimpleFinTokenSecretsParams{Kind: string(kind), Now: &nowUTC})
	return dbutil.MapRows(rows, err, simpleFinSecretFromRow)
}

func (s *Store) SimpleFinTokensDueForSync(ctx context.Context, now time.Time) ([]accounts.SimpleFinAccessTokenSecret, error) {
	return SimpleFinTokenSecretsDue(ctx, s.q, accounts.SimpleFinTokenSecretKindSync, now)
}

func (s *Store) SetSimpleFinTokenSynced(ctx context.Context, id int64, lastSyncedAt time.Time, nextSyncAt time.Time) error {
	last := lastSyncedAt.UTC()
	next := nextSyncAt.UTC()
	return s.q.SetSimpleFinTokenSynced(ctx, dbgen.SetSimpleFinTokenSyncedParams{
		ID:           id,
		LastSyncedAt: &last,
		NextSyncAt:   &next,
	})
}

func (s *Store) DeleteSimpleFinAccessToken(ctx context.Context, id int64) error {
	return s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		connections, err := q.SimpleFinConnections(ctx, dbgen.SimpleFinConnectionsParams{AccessTokenIds: []int64{id}})
		if err != nil {
			return fmt.Errorf("list simplefin connections: %w", err)
		}
		for _, conn := range connections {
			connectionID := conn.ConnectionID
			if err := q.DeleteAccountsByConnection(ctx, dbgen.DeleteAccountsByConnectionParams{ConnectionID: &connectionID}); err != nil {
				return fmt.Errorf("delete accounts: %w", err)
			}
			if err := q.DeleteConnectionByID(ctx, dbgen.DeleteConnectionByIDParams{ID: connectionID}); err != nil {
				return fmt.Errorf("delete connection: %w", err)
			}
			if err := q.DeleteSimpleFinConnection(ctx, dbgen.DeleteSimpleFinConnectionParams{ID: conn.ID}); err != nil {
				return fmt.Errorf("delete simplefin connection: %w", err)
			}
		}
		if err := q.DeleteSimpleFinAccessToken(ctx, dbgen.DeleteSimpleFinAccessTokenParams{ID: id}); err != nil {
			return fmt.Errorf("delete simplefin access token: %w", err)
		}
		return nil
	})
}

func (s *Store) ResetSimpleFinTokenSyncedAt(ctx context.Context, id int64) error {
	return s.q.ResetSimpleFinTokenSyncedAt(ctx, dbgen.ResetSimpleFinTokenSyncedAtParams{ID: id})
}

File: server-rs/src/accounts/store/simplefin_tokens.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    accounts::simplefin_types::SimpleFinAccessTokenSecret,
    accounts::{SimpleFinAccessToken, SimpleFinTokenSecretKind},
    database::{self, queries},
};

pub const DEFAULT_SIMPLEFIN_SYNC_CRON: &str = "0 6,18 * * *";

pub async fn create_simple_fin_access_token(
    pool: &SqlitePool,
    access_url: &str,
    owner_id: i64,
    label: &str,
) -> Result<SimpleFinAccessToken> {
    let id = queries::create_simple_fin_access_token(
        pool,
        queries::CreateSimpleFinAccessTokenParams {
            access_url,
            label: (!label.is_empty()).then_some(label),
            owner_id,
            sync_cron: DEFAULT_SIMPLEFIN_SYNC_CRON,
            next_sync_at: Some(Utc::now().into()),
        },
    )
    .await
    .context("create simplefin access token")?
    .id;
    simple_fin_access_token_by_id(pool, id)
        .await?
        .ok_or_else(|| anyhow::anyhow!("simplefin access token {id} not found"))
}

pub async fn simple_fin_access_tokens(
    executor: impl Executor<'_, Database = Sqlite>,
) -> Result<Vec<SimpleFinAccessToken>> {
    queries::simple_fin_access_tokens(executor, queries::SimpleFinAccessTokensParams::default())
        .await
        .map(|rows| rows.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}

pub async fn simple_fin_access_token_by_id(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
) -> Result<Option<SimpleFinAccessToken>> {
    queries::simple_fin_access_tokens(executor, queries::SimpleFinAccessTokensParams { id: Some(id) })
        .await
        .map(|rows| rows.into_iter().next().map(Into::into))
        .map_err(Into::into)
}

pub async fn simple_fin_token_secret_by_conn_id(
    executor: impl Executor<'_, Database = Sqlite>,
    conn_id: i64,
) -> Result<Option<SimpleFinAccessTokenSecret>> {
    queries::simple_fin_token_secret_by_conn_id_opt(executor, queries::SimpleFinTokenSecretByConnIdParams { conn_id })
        .await
        .map(|row| row.map(Into::into))
        .map_err(Into::into)
}

pub async fn simple_fin_token_secrets_due(
    executor: impl Executor<'_, Database = Sqlite>,
    kind: SimpleFinTokenSecretKind,
    now: DateTime<Utc>,
) -> Result<Vec<SimpleFinAccessTokenSecret>> {
    let kind = kind.to_string();
    queries::simple_fin_token_secrets(
        executor,
        queries::SimpleFinTokenSecretsParams {
            kind: &kind,
            now: Some(now.into()),
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn set_simple_fin_token_synced(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
    last_synced_at: DateTime<Utc>,
    next_sync_at: DateTime<Utc>,
) -> Result<()> {
    queries::set_simple_fin_token_synced(
        executor,
        queries::SetSimpleFinTokenSyncedParams {
            last_synced_at: Some(last_synced_at.into()),
            next_sync_at: Some(next_sync_at.into()),
            id,
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn set_simple_fin_token_balance_synced(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
    next_balance_sync_at: DateTime<Utc>,
) -> Result<()> {
    queries::set_simple_fin_token_balance_synced(
        executor,
        queries::SetSimpleFinTokenBalanceSyncedParams {
            next_balance_sync_at: Some(next_balance_sync_at.into()),
            id,
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn delete_simple_fin_access_token(pool: &SqlitePool, id: i64) -> Result<()> {
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let connections = queries::simple_fin_connections(
                &mut **transaction,
                queries::SimpleFinConnectionsParams {
                    access_token_ids: Some(&[id]),
                    ..Default::default()
                },
            )
            .await
            .context("list simplefin connections")?;
            for connection in connections {
                queries::delete_accounts_by_connection(
                    &mut **transaction,
                    queries::DeleteAccountsByConnectionParams {
                        connection_id: Some(connection.connection_id),
                    },
                )
                .await
                .context("delete accounts")?;
                queries::delete_connection_by_id(
                    &mut **transaction,
                    queries::DeleteConnectionByIdParams {
                        id: connection.connection_id,
                    },
                )
                .await
                .context("delete connection")?;
                queries::delete_simple_fin_connection(
                    &mut **transaction,
                    queries::DeleteSimpleFinConnectionParams { id: connection.id },
                )
                .await
                .context("delete simplefin connection")?;
            }
            queries::delete_simple_fin_access_token(
                &mut **transaction,
                queries::DeleteSimpleFinAccessTokenParams { id },
            )
            .await
            .context("delete simplefin access token")
        })
    })
    .await
}

pub async fn reset_simple_fin_token_synced_at(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<()> {
    queries::reset_simple_fin_token_synced_at(executor, queries::ResetSimpleFinTokenSyncedAtParams { id })
        .await
        .map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use anyhow::Result;
    use chrono::{DateTime, Utc};

    use crate::{
        accounts::SimpleFinAccessTokenSecret,
        database::{Timestamp, dbtest},
        testutil::store::create_owner,
    };

    use super::*;

    #[tokio::test]
    async fn reused_simplefin_token_secret_round_trips_storage_fields() -> Result<()> {
        let pool = dbtest::open().await?;
        let owner = create_owner(&pool, "alex").await?;
        let token = create_simple_fin_access_token(&pool, "https://bridge.example", owner.id, "Bridge").await?;
        let (simple_fin_connection_id, _) = super::super::link_simple_fin_connection(
            &pool,
            &super::super::test_support::simplefin_connection(token.id, owner.id, "simplefin"),
        )
        .await?;
        let timestamp = "2026-09-06T12:00:00Z".parse::<DateTime<Utc>>()?;
        sqlx::query(
            "UPDATE simplefin_access_tokens SET created_at = ?, updated_at = ?, last_synced_at = ?, next_sync_at = ?, next_balance_sync_at = ? WHERE id = ?",
        )
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(token.id)
        .execute(&pool)
        .await?;

        assert_eq!(
            simple_fin_token_secret_by_conn_id(&pool, simple_fin_connection_id)
                .await?
                .unwrap(),
            SimpleFinAccessTokenSecret {
                access_url: "https://bridge.example".to_owned(),
                created_at: timestamp.into(),
                id: token.id,
                label: Some("Bridge".to_owned()),
                last_synced_at: Some(timestamp.into()),
                next_balance_sync_at: Some(timestamp.into()),
                next_sync_at: Some(timestamp.into()),
                owner_id: owner.id,
                sync_cron: DEFAULT_SIMPLEFIN_SYNC_CRON.to_owned(),
                updated_at: timestamp.into(),
            }
        );
        Ok(())
    }
}
